controller/src/main/java/io/pravega/controller/task/Stream/StreamMetadataTasks.java

```diff
@@ -944,118 +944,158 @@ private boolean isSubscriberToNonSubscriberTransition(final ReaderGroupConfigRec
 
     private CompletableFuture<Map<Long, Long>> getTruncationStreamCutBySizeLimit(String scope, String stream, OperationContext context, RetentionPolicy policy,
                                                                                  RetentionSet retentionSet, Map<Long, Long> lowerBound) {
-        // if the lowerbound on subscribers streamcuts satisfies the policy size bound, then return it. 
-        // else return the stream cut that satisfies maximum bound on size. 
         // 1. if lowerbound.size < max and lowerbound.size > min truncate at lowerbound
-        // 2. if lowerbound.size < min, truncate at max irrespective of if lowerbound overlaps with max or not. 
+        // 2. if lowerbound.size < min, truncate at streamcut behind lowerbound that satisfies the policy. (we look at max and below).
         // 3. if lowerbound.size > max, truncate at max
         long currentSize = retentionSet.getLatest().getRecordingSize();
+        Map.Entry<StreamCutReferenceRecord, StreamCutReferenceRecord> limits = getBoundStreamCuts(policy, retentionSet,
+                x -> currentSize - x.getRecordingSize());
 
         // if lowerbound is empty simply return min
         if (lowerBound == null || lowerBound.isEmpty()) {
-            return retentionSet
-                    .getRetentionRecords().stream()
-                    .filter(x -> currentSize - x.getRecordingSize() > policy.getRetentionParam())
-                    .max(Comparator.comparingLong(StreamCutReferenceRecord::getRecordingTime))
+            return Optional.ofNullable(limits.getValue())
```

Comment:
This perhaps means if there is no subscriber.lowerbound, we truncate at a Stream-Cut that satisfies the `min`. What does `limits` have?

Reply:
limits has min and max.